Add module to generate Graph Embeddings

ShobhitBehl opened the pull request:
Added common/Vectorize.py to generate graph embeddings and compute pairwise cosine similarity of models in the spanner database.
Added a process_nodes() functionality in common/Graph.py to compute node reachability from input and delete unreachable nodes, Standardizes processing across TF and TFLite.
Updated README.md.

ShobhitBehl committed:
Add Basic Support for StatefulPartitionedCall

Added operator support for StatefulPartitionedCall in TF Parsing.
Added sub_category attribute in common/Graph.py.
Added is_canonical field in Models table, stored using an argument to
common/Storage.py, input using CLA.
Updated common/visualize.ipynb.

Update README.md

ShobhitBehl committed:
Add python code for loading data

Add load_data.py for loading data into the database.
Typo fixes in description of is_canonical field.
Added 'source' field extraction for graphs with StatefulPartitionedCall.

ShobhitBehl committed:
Add handling for no inputs found

Added conditional check for no input placholders found in
TF/src/TFParser.py.
Bug fix in common/Storage.py, was not using NOT_STORED_ATTR.

ShobhitBehl committed:
Add common/Vectorize.py

Added common/Vectorize module to create graph embeddings using models in
database using Graph2Vec.
Only the geature 'operator_type' considered for a node.
Added process_nodes() functionality to common/Graph.py to compute
reachability of nodes from input and discard if unreachable.
Added a call to process_nodes() from TFLite/TFLiteParser.py.
Bug fix in TF/main.py to return on finding None graph in load_data().
Added is_input boolena column in Operators Table to identify input operators,
made changes in common/Storage.py to load the same.

ShobhitBehl committed:
Add CLA for edge and node attributes in feature

Added class attributes to define the attribute names from Node and Edge
to be used when building feature for graph2vec.
Added CLA to optionally use Edge and Ndode attributes listed in the class
attributes.
Added binary build for common/Vectorize.py.
Updated README.md

ShobhitBehl committed:
Add common/similarity.py

Added common/similarity.py to run similarity of embeddings, removed the
same from common/Vectorize.py for it to be strictly an embedding module.
Modified build for 'Vectorize' into a library and added binary build for
'similarity'.
Added "None" category support in load_data.py
Updated README.md

ShobhitBehl committed:
Add time performance measures

Added time performance measure to check time to query database models
and create Graphs and time to fit graph2vec model.
Fixed a bug with interchanged values of include_node_attrs and
include_edge_attrs.
Added docstring for Graph2Vec parameters.
Updated common/visualize.ipynb.

prakashprabhu commented on common/Vectorize.py:
```diff
@@ -0,0 +1,307 @@
+        WORKERS (int) : Number of cores.
+        """      
+
+        WL_ITERATIONS = 3
+        ATTRIBUTED = True
+        EPOCHS = 500
+        LEARNING_RATE = 0.15
+        MIN_COUNT = 5
+        DIMENSIONS = 128
```
Do you have a sense for how the convergence/accuracy (for our task of clustering) varies with changes in these values? Also, a comment on how these values were chosen would be helpful (even they are arbitrary for now).

ShobhitBehl replied:
Added comment for the same.

prakashprabhu commented on common/Vectorize.py:
```diff
@@ -0,0 +1,307 @@
+
+        return networkx_graph
+
+    def get_graph2vec_embeddings(self, include_edge_attrs = False, 
```
Are the embeddings also stored in the db? I guess then when we add a new model, we can trigger a re-training automatically (using Spanner queues and receivers) to get new embeddings since the training is on the entire set of existing models - not immediately required, something to think about for future.

ShobhitBehl replied:
Added 'embeddings' field in Models table and added function to load the embeddings.

ShobhitBehl committed:
Add Embeddings to Models Table

Added embeddings storage in Models table in the database, added
functions for storing the same in common/Storage.py and
common/similarity.py.
Added wl_iterations as CLA.
Moved parse_models() functionality from common/Vectorize.py
to common/Storage.py.
Made signature changes to get_graph2vec_embeddings() in
common/Vectorize.py for it to take a list of model graphs as
argument instead of parsing models from database.

rachitagrawal commented on load_data.py:
```diff
@@ -0,0 +1,99 @@
+    "FaceDetection", "EmotionRecognition", "PalmDetection"
+    ]
+AUDIO_SUB_CATEGORIES = ["AutomaticSpeechRecognition"]
+
```
nit: un-necessary blank line. Same below as well.

ShobhitBehl replied:
Fixed

rachitagrawal commented on common/Storage.py:
```diff
@@ -314,4 +332,165 @@ def load_data(self, graph):
+
+        for row in results1:
+
+            if row[4] == 0 and not parse_stateful:
```
nit: A comment stating that you can checking num_inputs.

ShobhitBehl replied:
Added comment for the same.

rachitagrawal commented on common/Vectorize.py:
```diff
@@ -0,0 +1,167 @@
+            include_edge_attrs (str) : Case insensitive string to denote 
+                whether to include edge attributes of input edge from 
+                _EDGE_ATTRS in feature, if "true" then they are included.
+            include_node_attrs (str) : Case insensitive string to denote
```
type and description doesn't match.

ShobhitBehl replied:
Fixed.

rachitagrawal commented on load_data.py:
```diff
@@ -0,0 +1,99 @@
+""" Module to multiple files into the database
+
```
nit: to load multiple files?

ShobhitBehl replied:
Fixed typo.

rachitagrawal commented on load_data.py:
```diff
@@ -0,0 +1,99 @@
+Module loads multiple models into database from MODELS_DIR of tf and tflite.
+The file naming format should be of the form modelname_subcategory.tflite or
+modelname_subcategory_(SavedModel/FrozenGraph).pb.
+Can only load one type of "is_canonical" models at a time.
```
Do remember to change this if you are changing the meaning of this column ('is_canonical')

ShobhitBehl replied:
Changed field to "model_type" in accordance to semantics of use.

rachitagrawal commented on common/similarity.py:
```diff
@@ -0,0 +1,117 @@
+                    model_graphs[indices[rank]].model_name)
+
+def db_module_similarity(model_graphs, embeddings, topk):
+    """ Module to print topk similar models to each module in database
```
nit: (.) period missing.

rachitagrawal commented on common/similarity.py:
```diff
@@ -0,0 +1,117 @@
+import Vectorize
+
+def topk_similarity(model_graphs, embeddings, topk):
+    """ Function to print topk similar models to each model
```
nit: (.) period missing.

rachitagrawal replied:
Same at many other places.

rachitagrawal commented on common/similarity.py:
```diff
@@ -0,0 +1,117 @@
+""" module to to display cosine similarity
```
nit: Capitalization and (.) missing.

ShobhitBehl replied:
Fixed similar problem across all files.

rachitagrawal commented on common/Vectorize.py:
```diff
@@ -0,0 +1,167 @@
+    """
+
+    _EDGE_ATTRS = [
+        "tensor_shape", "tensor_type"
```
Are there any limitation for edge attributes and node attributes? If yes, can you add comment about the same here.

ShobhitBehl replied:
Can only be attributes of Node/Edge class, added a comment for the same.

rachitagrawal commented on common/Storage.py:
```diff
@@ -314,4 +332,165 @@ def load_data(self, graph):
+                " FROM Models"
+                )
+
+        for row in results1:
```
nit: 'results1' --> 'qresult_models'?

ShobhitBehl replied:
Changed naming to 'qresult_models'.

rachitagrawal commented on common/Storage.py:
```diff
@@ -314,4 +332,165 @@ def load_data(self, graph):
+
+            # Querying Operators of model_name
+            with self.database.snapshot() as snapshot:
+                results2 = snapshot.execute_sql(
```
Same here. may be qresult_operators?

ShobhitBehl replied:
Changed naming to 'qresult_operators'.

rachitagrawal commented on common/Storage.py:
```diff
@@ -314,4 +332,165 @@ def load_data(self, graph):
+            
+            # Querying Tensors of model_name
+            with self.database.snapshot() as snapshot:
+                results2 = snapshot.execute_sql(
```
more descriptive name.

ShobhitBehl replied:
Changed naming to 'qresult_tensors'.

rachitagrawal reviewed (commented):
Possibly reviewed all the files in this PR.

ShobhitBehl committed:
Changed "is_canonical" field to "model_type"

Changes the is_canonical field in DB to "model_type" for extended use,
made changes in files for the same.
Fixed comments.
Fixed query result naming in common/Storage.py
Fixed bug in common/Vectorize.py of arguments defaulted to wrong type.

rachitagrawal reviewed (approved):
LGTM
